docs/source/reference/build_config.rst

```diff
@@ -0,0 +1,357 @@
+============
+Build Config
+============
+
+Build Config are used to represent the structure of
+a :code:`bentofile.yaml` in our Python API. 
+
+.. warning::
+
+    This page contains BentoML internal API reference. This documentation is
+    provided for developers who wants to explore BentoML internals as well as
+    extends BentoML features. 
+
+.. note::
+
+    Refers to :ref:`concepts/bento:Building Bentos` for more information about how to build a bento.
+
+BentoML under the hood utilize `attrs library <https://www.attrs.org/en/stable/index.html>`_ 
+to provide a better experience writing classes in Python. They comes with features, such as
+validation, conversion, supports for type-checking, etc. 
+
+All classes below are :code:`attrs.frozen()` class, meaning values are immutable to change.
+Refers to `attrs.frozen() API reference <https://www.attrs.org/en/stable/api.html#attr.attrs.frozen>`_.
+
+
+BentoBuildConfig
+----------------
+
+.. autoclass:: bentoml._internal.bento.build_config.BentoBuildConfig
+    :members:
+
+
+PythonOptions
+-------------
+
+.. autoclass:: bentoml._internal.bento.build_config.PythonOptions
+    :members:
+
+
+CondaOptions
+------------
+
+.. autoclass:: bentoml._internal.bento.build_config.CondaOptions
+    :members:
+
+
+DockerOptions
+-------------
+
+.. autoclass:: bentoml._internal.bento.build_config.DockerOptions
+    :members:
+
+Dockerfile generation
```

Comment:
could you open a new PR with just the "Dockerfile generation" part in a advanced guide page?